nixos/modules/virtualisation/cri-o.nix

```diff
@@ -80,60 +85,67 @@ in
-
-      [crio.network]
-      plugin_dirs = ["${pkgs.cni-plugins}/bin/"]
-
-      [crio.runtime]
-      cgroup_manager = "systemd"
-      log_level = "${cfg.logLevel}"
-      manage_ns_lifecycle = true
-      pinns_path = "${cfg.package}/bin/pinns"
-      hooks_dir = []
-
-      ${optionalString (cfg.runtime != null) ''
-      default_runtime = "${cfg.runtime}"
-      [crio.runtime.runtimes]
-      [crio.runtime.runtimes.${cfg.runtime}]
-      ''}
-    '';
-
-    environment.etc."cni/net.d/10-crio-bridge.conf".source = copyFile "${pkgs.cri-o-unwrapped.src}/contrib/cni/10-crio-bridge.conf";
-    environment.etc."cni/net.d/99-loopback.conf".source = copyFile "${pkgs.cri-o-unwrapped.src}/contrib/cni/99-loopback.conf";
-
-    # Enable common /etc/containers configuration
-    virtualisation.containers.enable = true;
-
-    systemd.services.crio = {
-      description = "Container Runtime Interface for OCI (CRI-O)";
-      documentation = [ "https://github.com/cri-o/cri-o" ];
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      path = [ cfg.package ];
-      serviceConfig = {
-        Type = "notify";
-        ExecStart = "${cfg.package}/bin/crio";
-        ExecReload = "/bin/kill -s HUP $MAINPID";
-        TasksMax = "infinity";
-        LimitNOFILE = "1048576";
-        LimitNPROC = "1048576";
-        LimitCORE = "infinity";
-        OOMScoreAdjust = "-999";
-        TimeoutStartSec = "0";
-        Restart = "on-abnormal";
+  config = mkMerge [
+    (mkIf cfg.enable {
+      environment.systemPackages = [ cfg.package pkgs.cri-tools ];
+
+      environment.etc."crictl.yaml".source = copyFile "${pkgs.cri-o-unwrapped.src}/crictl.yaml";
+
+      environment.etc."crio/crio.conf.d/00-default.conf".text = ''
+        [crio]
+        storage_driver = "${cfg.storageDriver}"
+
+        [crio.image]
+        ${optionalString (cfg.pauseImage != null) ''pause_image = "${cfg.pauseImage}"''}
+        ${optionalString (cfg.pauseCommand != null) ''pause_command = "${cfg.pauseCommand}"''}
+
+        [crio.network]
+        plugin_dirs = ["${pkgs.cni-plugins}/bin/"]
+        ${optionalString (cfg.networkDir != null) ''network_dir = "${cfg.networkDir}"''}
+
+        [crio.runtime]
+        cgroup_manager = "systemd"
+        log_level = "${cfg.logLevel}"
+        manage_ns_lifecycle = true
+        pinns_path = "${cfg.package}/bin/pinns"
+        hooks_dir = []
+
+        ${optionalString (cfg.runtime != null) ''
+        default_runtime = "${cfg.runtime}"
+        [crio.runtime.runtimes]
+        [crio.runtime.runtimes.${cfg.runtime}]
+        ''}
+      '';
+
+      # Enable common /etc/containers configuration
+      virtualisation.containers.enable = true;
+
+      # Loopback should be always added independently if a custom network dir is used or not
+      environment.etc."cni/net.d/99-loopback.conf".source = copyFile "${pkgs.cri-o-unwrapped.src}/contrib/cni/99-loopback.conf";
```

Comment:
Referencing to https://github.com/NixOS/nixpkgs/pull/98086/files#r493229379

Do you see any good way how we can make this point to `/etc/cni/net.d` by default and if `cfg.networkDir != null` to the target location?

I'd tried it with something like `environment."${cfg.networkDir}/99-loopback.conf".source` but this would not work at all.

Reply:
We could use `networkDir` as an internal option that is only set by other modules and expose a `cni_default_network` option to users to let them choose a non-default config in `/etc/cni/net.d` if `networkDir` isn't already set by another module?

Reply:
Hm, how would this look like? I now have the `networkDir` set in the config, but how to make the option only internally available?

Reply:
This should work:

```nix
    networkDir = mkOption {
      type = types.nullOr types.path;
      default = null;
      description = "...";
      internal = true;
    };
```

Reply:
Indeed, thank you! :pray: Changed as suggested.